url70ed135da7_ralfbiedert_jcores_tgz-pJ8-sandbox_SimpleFilterTestJ8/src/net/jcores/jre/utils/internal/Files.java

```diff
@@ -0,0 +1,317 @@
+
+            while ((line = reader.readLine()) != null) {
+                sb.append(line);
+                sb.append("\n");
+            }
+
+            return sb.toString();
+        } catch (FileNotFoundException e) {
+            cc.report(MessageType.EXCEPTION, "File not found " + file);
+        } catch (IOException e) {
+            cc.report(MessageType.EXCEPTION, "Error reading from file " + file);
+        } finally {
+            if (reader != null) try {
+                reader.close();
+            } catch (IOException e) {
+                cc.report(MessageType.EXCEPTION, "Error closing file " + file);
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Lists all elements under the given root.
+     * 
+     * @param root
+     * @param listDirs
+     * @return A list of elements
+     */
+    public static File[] dir(File root, boolean listDirs) {
+        final List<File> rval = new ArrayList<File>();
+
+        // Get top level files ...
+        List<File> next = new ArrayList<File>();
+        File[] listed = root.listFiles();
+
+        // In case there are no files in the given diretory element, return
+        if (listed == null) return null;
+
+        // Queue all we found
+        next.addAll(Arrays.asList(listed));
+
+        while (next.size() > 0) {
+            // Take over all from the next queue
+            listed = next.toArray(new File[0]);
+            next.clear();
+
+            // Now check for each item
+            for (File file : listed) {
+                if (!file.isDirectory()) {
+                    rval.add(file);
+                    continue;
+                }
+
+                if (listDirs) rval.add(file);
+
+                final File[] listFiles = file.listFiles();
+                if (listFiles == null) continue;
+                next.addAll(Arrays.asList(listFiles));
+            }
+        }
+
+        return rval.toArray(new File[0]);
+    }
+
+    /**
+     * Copies a file .
+     * 
+     * @param cc
+     * @param from
+     * @param to
+     * @return .
+     */
+    public static File[] copy(CommonCore cc, File from, File to) {
+
+        // Create directory if they don't exist
+        final boolean todir = to.getAbsolutePath().endsWith("/") || to.isDirectory();
+        final boolean fromdir = from.getAbsolutePath().endsWith("/") || from.isDirectory();
+
+        // If we had a source dir
+        if (fromdir) {
+            final File[] elements = dir(from, false);
+            final List<File> files = new ArrayList<File>();
+            for (File file : elements) {
+                final String subname = file.getAbsolutePath().replace(from.getAbsolutePath(), "");
+                files.addAll(Arrays.asList(copy(cc, file, new File(to + "/" + subname))));
+            }
+
+            return files.toArray(new File[0]);
+        }
+
+        // If its a dir, create the dir, if its a file, create its parent
+        if (todir) to.mkdirs();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Confirm that `mkdirs` was successful. Handle any failures that happened. If you fail silently, then you might encounter errors that are harder to debug. See the following diffs on GitHub that make this code change: [diff1](https://github.com/qunarcorp/qmq/commit/12f1d9d6a57a02bc2415cb2af9f4388a77b12c02#diff-6b67be6e17f24fdeddb7c725ac8ec5600d6e63be0bdca0963a83d530e6d55e35L36), [diff2](https://github.com/apache/jmeter/commit/1ec61206779b3f84df18516682b4791bf015178f#diff-cd74304763c097fbc085ac2bc952b46ab3843ac825cacacad89e8064296e955aL330)

Learn more: [File.mkdirs in the Java Platform, Standard Edition 8 API Specification](https://docs.oracle.com/javase/8/docs/api/java/io/File.html#mkdirs--)